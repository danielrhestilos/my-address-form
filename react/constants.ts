import type { PostalCodeSource } from './types/rules'

export const POSTAL_CODE = 'POSTAL_CODE' as const
export const ONE_LEVEL = 'ONE_LEVEL' as const
export const TWO_LEVELS = 'TWO_LEVELS' as const
export const THREE_LEVELS = 'THREE_LEVELS' as const

export const POSTAL_CODE_FROM_TYPES = [
  POSTAL_CODE,
  ONE_LEVEL,
  TWO_LEVELS,
  THREE_LEVELS,
] as PostalCodeSource[]

export const EEMPTY = 'ERROR_EMPTY_FIELD'
export const EADDRESSTYPE = 'ERROR_ADDRESS_TYPE'
export const ENOTOPTION = 'ERROR_VALUE_IS_NOT_AN_OPTION'
export const ECOUNTRY = 'ERROR_COUNTRY_CODE'
export const EGEOCOORDS = 'ERROR_GEO_COORDS'
export const EPOSTALCODE = 'ERROR_POSTAL_CODE'
export const EGOOGLEADDRESS = 'ERROR_GOOGLE_ADDRESS'

export const ERROR_TYPES = [
  EEMPTY,
  EADDRESSTYPE,
  ENOTOPTION,
  ECOUNTRY,
  EGEOCOORDS,
  EPOSTALCODE,
  EGOOGLEADDRESS,
]
