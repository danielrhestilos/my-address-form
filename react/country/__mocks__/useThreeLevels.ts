import { THREE_LEVELS } from '../../constants'
import { thirdLevelPostalCodes } from '../../transforms/postalCodes'
import {
  getOneLevel,
  getTwoLevels,
  getThreeLevels,
} from '../../transforms/addressFieldsOptions'
import type { PostalCodeRules } from '../../types/rules'

const countryData = {
  Beni: {
    Cercado: {
      Paititi: '10000',
      Trinidad: '10001',
      Yucumo: '10002',
    },
    Mamore: {
      'San Ramon': '10200',
    },
  },
  Cochabamba: {
    Arani: {
      Arani: '30000',
      Vacas: '30001',
    },
    Chapare: {
      Colomi: '30500',
      Epizana: '30501',
      Eterazama: '30502',
    },
  },
}

const rules: PostalCodeRules = {
  country: 'BOL',
  abbr: 'BO',
  postalCodeFrom: THREE_LEVELS,
  postalCodeLevels: ['state', 'city', 'neighborhood'],
  thirdLevelPostalCodes: thirdLevelPostalCodes(countryData),
  fields: [
    {
      name: 'street',
      label: 'street',
      required: true,
      size: 'xlarge',
    },
    {
      name: 'number',
      maxLength: 750,
      label: 'number',
      required: true,
      size: 'mini',
    },
    {
      name: 'complement',
      maxLength: 750,
      label: 'complement',
      size: 'large',
    },
    {
      name: 'state',
      maxLength: 100,
      label: 'department',
      required: true,
      size: 'large',
      level: 1,
      options: getOneLevel(countryData),
    },
    {
      name: 'city',
      label: 'province',
      required: true,
      size: 'large',
      level: 2,
      basedOn: 'state',
      optionsMap: getTwoLevels(countryData),
    },
    {
      name: 'neighborhood',
      maxLength: 100,
      label: 'city',
      required: true,
      level: 3,
      basedOn: 'city',
      optionsMap: getThreeLevels(countryData),
    },
    {
      name: 'receiverName',
      elementName: 'receiver',
      maxLength: 750,
      label: 'receiverName',
      size: 'large',
      required: true,
    },
  ],
}

export default rules
