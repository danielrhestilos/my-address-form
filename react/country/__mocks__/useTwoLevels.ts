import { TWO_LEVELS } from '../../constants'
import { secondLevelPostalCodes } from '../../transforms/postalCodes'
import {
  getOneLevel,
  getTwoLevels,
} from '../../transforms/addressFieldsOptions'
import type { PostalCodeRules } from '../../types/rules'

const countryData = {
  'I Región': {
    Camiña: '1150000',
    Colchane: '1160000',
    Huara: '1140000',
    Iquique: '1100000',
    Pica: '1170000',
  },
  'II Región': {
    Antofagasta: '1240000',
    Calama: '1390000',
  },
}

const rules: PostalCodeRules = {
  country: 'CHL',
  abbr: 'CL',
  postalCodeFrom: TWO_LEVELS,
  postalCodeLevels: ['state', 'neighborhood'],
  secondLevelPostalCodes: secondLevelPostalCodes(countryData),
  fields: [
    {
      name: 'street',
      label: 'street',
      required: true,
      size: 'xlarge',
    },
    {
      name: 'number',
      maxLength: 750,
      label: 'number',
      required: true,
      size: 'mini',
    },
    {
      name: 'complement',
      maxLength: 750,
      label: 'complement',
      size: 'large',
    },
    {
      name: 'state',
      maxLength: 100,
      label: 'region',
      required: true,
      size: 'large',
      options: getOneLevel(countryData),
    },
    {
      name: 'neighborhood',
      maxLength: 100,
      label: 'community',
      required: true,
      size: 'large',
      level: 2,
      basedOn: 'state',
      optionsMap: getTwoLevels(countryData),
    },
    {
      name: 'receiverName',
      elementName: 'receiver',
      maxLength: 750,
      label: 'receiverName',
      size: 'large',
      required: true,
    },
  ],
}

export default rules
