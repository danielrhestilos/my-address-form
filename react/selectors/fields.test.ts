import diff from 'lodash/difference'

import {
  getField,
  hasOptions,
  getListOfOptions,
  getDependentFields,
  filterPostalCodeFields,
  filterAutoCompletedFields,
  isDefiningPostalCodeField,
  getFieldLabel,
} from './fields'
import { ONE_LEVEL, TWO_LEVELS, POSTAL_CODE } from '../constants'
import useOneLevel from '../country/__mocks__/useOneLevel'
import useTwoLevels from '../country/__mocks__/useTwoLevels'
import useThreeLevels from '../country/__mocks__/useThreeLevels'
import type { PostalCodeRules, PostalCodeFieldRule } from '../types/rules'
import type { AddressWithValidation } from '../types/address'

describe('Field Selectors', () => {
  it('getField()', () => {
    const rules: PostalCodeRules = {
      country: null,
      abbr: null,
      fields: [
        { name: 'postalCode', label: 'zipCode' },
        { name: 'city', label: 'community' },
      ],
    }

    const field = getField('city', rules) as PostalCodeFieldRule

    expect(getFieldLabel(field)).toBe('community')
  })

  describe('hasOptions()', () => {
    it('for field with options', () => {
      const field: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        options: ['PE', 'SP'],
      }

      const result = hasOptions(field)

      expect(result).toBe(true)
    })

    it('for field with optionsPairs', () => {
      const field: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        optionsPairs: [
          { value: 'PE', label: 'Pernambuco' },
          { value: 'SP', label: 'São Paulo' },
        ],
      }

      const result = hasOptions(field)

      expect(result).toBe(true)
    })

    it('for field with optionsMap', () => {
      const field: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        optionsMap: {
          PE: ['Recife', 'Olinda'],
          SP: ['São Paulo', 'Santos'],
        },
      }

      const result = hasOptions(field)

      expect(result).toBe(true)
    })
  })

  describe('getListOfOptions()', () => {
    it('field with options should get an object of value and label', () => {
      const field: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        options: ['PE', 'SP'],
      }

      const address = {} as AddressWithValidation
      const rules = {} as PostalCodeRules

      const options = getListOfOptions(field, address, rules)

      expect(options[0]).toMatchObject({ value: 'PE', label: 'PE' })
    })

    it('field with optionsPairs should return its optionsPairs', () => {
      const field: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        optionsPairs: [
          { value: 'PE', label: 'Pernambuco' },
          { value: 'SP', label: 'São Paulo' },
        ],
      }

      const address = {} as AddressWithValidation
      const rules = {} as PostalCodeRules

      const options = getListOfOptions(field, address, rules)

      expect(options[0]).toMatchObject({ value: 'PE', label: 'Pernambuco' })
    })

    it('field with optionsMap level 2 should get options based on other field value', () => {
      const field: PostalCodeFieldRule = {
        name: 'city',
        label: 'city',
        basedOn: 'state',
        level: 2,
        optionsMap: {
          Antioquia: ['Barranco Minas', 'Cacahual'],
        },
      }

      const address = { state: { value: 'ANTIOQUIA' } } as AddressWithValidation
      const rules = {} as PostalCodeRules

      const options = getListOfOptions(field, address, rules)

      expect(options[0]).toMatchObject({
        value: 'Barranco Minas',
        label: 'Barranco Minas',
      })
    })

    it('field with optionsMap level 3 should get options based on other field value', () => {
      const field: PostalCodeFieldRule = {
        name: 'neighborhood',
        label: 'neighborhood',
        basedOn: 'city',
        level: 3,
        optionsMap: {
          PE: {
            Recife: ['Boa Viagem', 'Casa Forte'],
          },
        },
      }

      const address = {
        state: { value: 'PE' },
        city: { value: 'Recife' },
      } as AddressWithValidation

      const rules = {
        postalCodeFrom: POSTAL_CODE,
        fields: [{ name: 'city', basedOn: 'state' }, { name: 'state' }],
      } as PostalCodeRules

      const options = getListOfOptions(field, address, rules)

      expect(options[0]).toMatchObject({
        value: 'Boa Viagem',
        label: 'Boa Viagem',
      })
    })

    it('should filter options suggested by postal code service', () => {
      const stateField: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        options: ['Foo', 'Bar', 'Zoo'],
      }

      const address = {
        state: { value: null, valueOptions: ['Bar', 'Too'] },
      } as AddressWithValidation

      const rules = {
        postalCodeFrom: POSTAL_CODE,
        fields: [stateField],
      } as PostalCodeRules

      const options = getListOfOptions(stateField, address, rules)

      expect(options).toHaveLength(1)
      expect(options[0]).toMatchObject({
        value: 'Bar',
        label: 'Bar',
      })
    })

    it('should fix options suggested by postal code service', () => {
      const stateField: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        options: ['Foo', 'Bar', 'Zóo'],
      }

      const address = {
        state: { value: null, valueOptions: ['Bar', 'zoo'] },
      } as AddressWithValidation

      const rules = {
        postalCodeFrom: POSTAL_CODE,
        fields: [stateField],
      } as PostalCodeRules

      const options = getListOfOptions(stateField, address, rules)

      expect(options).toHaveLength(2)
      expect(options[1]).toMatchObject({
        value: 'Zóo',
        label: 'Zóo',
      })
    })

    it('should not display options if basedOn field is not filled', () => {
      const stateField: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        options: ['Foo', 'Bar', 'Zóo'],
      }

      const cityField: PostalCodeFieldRule = {
        name: 'city',
        label: 'city',
        basedOn: 'state',
        level: 2,
        optionsMap: { Foo: ['Foolite'], Bar: ['Bartoo'] },
      }

      const address = {
        state: { value: null, valueOptions: ['Bar', 'zoo'] },
        city: { value: null, valueOptions: ['Foolite', 'Bartoo'] },
      } as AddressWithValidation

      const rules = {
        postalCodeFrom: POSTAL_CODE,
        fields: [stateField, cityField],
      } as PostalCodeRules

      const options = getListOfOptions(cityField, address, rules)

      expect(options).toHaveLength(0)
    })

    it('should display only the suggested options of the basedOn field', () => {
      const stateField: PostalCodeFieldRule = {
        name: 'state',
        label: 'state',
        options: ['Foo', 'Bar', 'Zóo'],
      }

      const cityField: PostalCodeFieldRule = {
        name: 'city',
        label: 'city',
        basedOn: 'state',
        level: 2,
        optionsMap: { Foo: ['Foolite'], Bar: ['Bartoo'] },
      }

      const address = {
        state: { value: 'Bar', valueOptions: ['Bar', 'zoo'] },
        city: { value: null, valueOptions: ['Foolite', 'Bartoo'] },
      } as AddressWithValidation

      const rules = {
        postalCodeFrom: POSTAL_CODE,
        fields: [stateField, cityField],
      } as PostalCodeRules

      const options = getListOfOptions(cityField, address, rules)

      expect(options).toHaveLength(1)
    })
  })

  describe('getDependentFields()', () => {
    it('with one level', () => {
      const rules = {
        postalCodeFrom: POSTAL_CODE,
        fields: [{ basedOn: 'state', name: 'city' }],
      } as PostalCodeRules

      const dependentFields = getDependentFields('state', rules)

      expect(dependentFields).toHaveLength(1)
      expect(dependentFields).toMatchObject(['city'])
    })

    it('with two levels', () => {
      const rules = {
        postalCodeFrom: POSTAL_CODE,
        fields: [
          { basedOn: 'state', name: 'city' },
          { basedOn: 'city', name: 'neighborhood' },
        ],
      } as PostalCodeRules

      const dependentFields = getDependentFields('state', rules)

      expect(dependentFields).toHaveLength(2)
      expect(dependentFields).toMatchObject(['city', 'neighborhood'])
    })

    it('postal code based on a field', () => {
      const rules = {
        country: null,
        abbr: null,
        postalCodeFrom: ONE_LEVEL,
        postalCodeLevels: ['state'],
        fields: [],
      } as PostalCodeRules

      const dependentFields = getDependentFields('state', rules)

      expect(dependentFields).toHaveLength(1)
      expect(dependentFields).toMatchObject(['postalCode'])
    })

    it("should not clear any if it's a field that doesn't define a postal code", () => {
      const rules = {
        country: null,
        abbr: null,
        postalCodeFrom: TWO_LEVELS,
        postalCodeLevels: ['state', 'city'],
        fields: [],
      } as PostalCodeRules

      const dependentFields = getDependentFields('street', rules)

      expect(dependentFields).toHaveLength(0)
    })
  })

  describe('filterPostalCodeFields()', () => {
    function getFieldNames(fields) {
      return fields.map(({ name }) => name)
    }

    it('should filter when postal code is from postal code', () => {
      const fields = filterPostalCodeFields(useOneLevel)

      expect(
        diff(getFieldNames(useOneLevel.fields), getFieldNames(fields))
      ).toMatchSnapshot()
    })

    it('should filter when postal code is from state', () => {
      const fields = filterPostalCodeFields(useTwoLevels)

      expect(
        diff(getFieldNames(useTwoLevels.fields), getFieldNames(fields))
      ).toMatchSnapshot()
    })

    it('should filter when postal code is from city', () => {
      const fields = filterPostalCodeFields(useThreeLevels)

      expect(
        diff(getFieldNames(useThreeLevels.fields), getFieldNames(fields))
      ).toMatchSnapshot()
    })
  })

  describe('isDefiningPostalCodeField()', () => {
    it('when rules have postalCodeLevels', () => {
      const rules: PostalCodeRules = {
        country: null,
        abbr: null,
        postalCodeLevels: ['state', 'city', 'neighborhood'],
        fields: [],
      }

      const result = isDefiningPostalCodeField('neighborhood', rules)

      expect(result).toBe(true)
    })
  })

  describe('filterAutoCompletedFields()', () => {
    it('should filter auto completed fields', () => {
      const address = ({
        neighborhood: { value: 'Botafogo', geolocationAutoCompleted: true },
        city: { value: 'Rio de Janeiro', postalCodeAutoCompleted: true },
        state: { value: 'RJ' },
      } as unknown) as AddressWithValidation

      const rules = {
        fields: [{ name: 'neighborhood' }, { name: 'city' }, { name: 'state' }],
      } as PostalCodeRules

      const result = filterAutoCompletedFields(rules, address)

      expect(result).toHaveLength(1)
      expect(result[0].name).toBe('state')
    })
  })
})
