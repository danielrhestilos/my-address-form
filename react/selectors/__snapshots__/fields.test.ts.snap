// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Selectors filterPostalCodeFields() should filter when postal code is from city 1`] = `
Array [
  "state",
  "city",
  "neighborhood",
]
`;

exports[`Field Selectors filterPostalCodeFields() should filter when postal code is from postal code 1`] = `
Array [
  "state",
]
`;

exports[`Field Selectors filterPostalCodeFields() should filter when postal code is from state 1`] = `
Array [
  "state",
  "neighborhood",
]
`;
